pkg/info/infopb/rpc.proto

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+syntax = "proto3";
+package thanos;
+
+import "store/labelpb/types.proto";
+import "gogoproto/gogo.proto";
+
+option go_package = "infopb";
+
+option (gogoproto.sizer_all) = true;
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+option (gogoproto.goproto_getters_all) = false;
+
+// Do not generate XXX fields to reduce memory footprint and opening a door
+// for zero-copy casts to/from prometheus data types.
+option (gogoproto.goproto_unkeyed_all) = false;
+option (gogoproto.goproto_unrecognized_all) = false;
+option (gogoproto.goproto_sizecache_all) = false;
+
+// Info represents the API that is responsible for gathering metadata about the all APIs supported by the component.
+service Info {
+    // Info returns the metadata (Eg. LabelSets, Min/Max time) about all the APIs the component supports.
+    rpc Info(InfoRequest) returns (InfoResponse);
+}
+
+message InfoRequest {}
+
+enum ComponentType {
+    UNKNOWN = 0;
+    QUERY = 1;
+    RULE = 2;
+    SIDECAR = 3;
+    STORE = 4;
+    RECEIVE = 5;
+    // DEBUG represents some debug StoreAPI components e.g. thanos tools store-api-serve.
+    DEBUG = 6;
+  }
+
+message InfoResponse {
+    repeated ZLabelSet label_sets = 1 [(gogoproto.nullable) = false];
+    ComponentType ComponentType  = 2;
```

Comment:
I expected this not to be the component type, but rather the APIs that this server serves, the type can be ambiguous, especially as we already have community extensions that don't strictly fall under any component, eg. the OpenTSDB implementation of the StoreAPI.

Reply:
We can discover APIs supported by the server using gRPC reflection and retrieve metadata by calling the `Info` method of this service. I have added `ComponentType` to be more consistent with the [Info](https://github.com/thanos-io/thanos/blob/7d118f416be98cf502b9b5e0df4e98b37478f3f3/pkg/store/storepb/rpc.proto#L79) method we already have in StoreAPI.
TBH, not sure about the compatibility of this Info service with community extensions.

Reply:
Yea, I think it's a misunderstanding a bit. The component type is really for debugging purposes, we (try) to not rely on this information. Maybe it should be then string in next iteration to indicate that it should not be used for discoverability? We use it for the Store UI page and some compatibility changes around metadata / label advertising in store GW. 🤔

Reply:
Just to clarify a bit more on why we plan to use reflection for discovery. If we use reflection we will have a single source of truth. For example, if a gRPC endpoint exposes `thanos.StoreAPI` only then the Querier will add it to the stores list and otherwise it won't. Another advantage that builds up on this is that if we add some Service to a component, it will be discovered automatically by the Querier without additional changes. For example, if we add support for exemplarsAPI in Receive, we will have to just do that (implement the API) and the Querier will pick this up automatically and add the Receive endpoint to `exemplars` list.

@brancz @bwplotka I am interested in your opinions on this. Do you think we should go with this approach or do you prefer exposing this information as part of the Info Service and not using reflection.

Reply:
I see, thanks for clarifying everyone. I think what @bwplotka said, to just change the component type to be a string instead of an enum would be best then, that way it can serve the debugging purpose without being constraining on the ecosystem.